### DISCUSSION TITLE
Price timeseries fetching for registered user portfolios

### DISCUSSION THREAD

**Discussion by leonardoarcari** [leonardoarcari/dcapal#441] [Ideas]
## Scope

The upcoming private area and portfolio sync features will enable users to create their portfolios on one device and access them on another one. Portfolios are stored anonymously in the backend Postgres and users will be able to pick up from where they left on one device, greatly improving the overall user experience.

The availability of portfolio allocations on the backend enables us to offer more powerful features. One of the biggest features missing are _historical performance of user's asset allocation_ (model portfolio) and _periodic PnL_ based on assets' Average Buy Price (ABP).  In order to implement those analytics, we need a common core feature: **price timeseries fetching**.

## Requirements

Given the inherent complexity on managing possibly long timeseries for the universe of investable assets, we need to consider some constraints to guide us in searching for an effective, yet scalable implementation.

1. Timeseries for an asset must be fetched once and kept up to date daily via last prices
2. Given the existing rate limits on our price providers, timeseries must be fetched in a batched, background fashion: if a user adds an asset to their portfolio and we don't have the timeseries in the DB, a job to fetch should be posted and we should inform the user that the timeseries will be displayed soon
3. In order to speed-up the timeseries retrieval, the client must present the desired period (e.g. daily/weekly/monthly) and the backend should fetch data from a pre-computed sampling of the requested period
4. Missing points in the timeseries will be filled client-side with a back-filling strategy
5. A periodic task should be scheduled in the backend to look for missing timeseries (e.g. an error occurred while fetching) or missing data points in the last N days
6. Unregistered users may be provided with current portfolio PnL if we have timeseries for all their assets. We _may_ consider to post a timeseries fetching job if the unregistered user completes the allocation process (i.e. it's a real portfolio, not a random test)

## Implementation strategy

In this section I wanna propose an implementation strategy to effectively implement the requirements above

- Add new `TimeseriesFetcher` worker in the backend, executing fetching jobs posted to a queue
- Add new `TimeseriesMaintenance` worker periodically checking for missing timeseries in a batch of portfolios and posting fetching jobs to `TimeseriesFetcher` queue
- Timeseries should be stored in [`timescaledb`](https://github.com/timescale/timescaledb), a powerful PosgreSQL extension providing timeseries capabilities to PostgresSQL (which we already use for Supabase)
- Timescale's [continuous aggregation](https://docs.timescale.com/getting-started/latest/aggregation/) feature provides automatic sampling of base timeseries (e.g. daily prices) into aggregated views (e.g. weekly/monthly prices)
- Add new endpoints to fetch assets timeseries provided:
  - `ticker`
  - `timeframe`
  - `from` and `to` dates
- The MVP for this feature might show portfolio model performance vs common benchmarks like S&P500, Nasdaq or VWCE
- Historical FX timeseries must be pre-fetched to support multi-currency portfolios

**Comment by corradopetrelli** [comment]
It all makes sense to me. The only thing: different tickers are used for different stock exchanges, so it is better to keep the concept of ticker connected to the reference stock exchange.

For example https://www.justetf.com/it/etf-profile.html?isin=IE00B4L5Y983#quotazione SWDA is called that on some stock exchanges, on others not.

**Reply by leonardoarcari** [reply]
That's a good point. As of today, tickers are strongly bound to the market data provider: crypto symbols are provided by Kraken, while all other assets are provided by Yahoo Finance and that's the ticker we are storing in Redis. I agree that if in the future we're gonna support different data providers for the same asset class, some sort of conversion table mapping our canonical ticker with the data provider's one.

**Comment by leonardoarcari** [comment]
Adding some more comments by @mattiapitossi as discussed offline:

> 1. **Portfolio anonymity** 
> 
> This is an important topic to address to guarantee users privacy. Some viable approaches include:
> 
>   - Use a random identifier instead of user ID to map portfolios to users.
>   - Encrypt personal information before we store it.
>   - Data segregation: store user information separate from portfolio data.
> 
> 2. **Pre-load timeseries for most popular assets**
> 
> This is a good opportunity to improve first-time user experience
> 
>   - Keep a list of the 50/100 most popular ETFs according to platforms like JustETF.
>   - Pre-load timeseries from the last 5/10 years
> 
> 3. **Misc**
> 
> Some more ideas to consider:
> 
>   - Implement a caching system to improve latency when accessing frequently accessed timeseries
>   - Implement a process to periodically keep assets up-to-date
